feat: allow vault secret to handle write operation

loispostula opened the pull request:
<!-- Thank you for contributing!

     Read more about how you can contribute in our contribution guide:
     https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md
-->

Allow hashicorp vault secret to use write operation

### Checklist

- [x] When introducing a new scaler, I agree with the [scaling governance policy](https://github.com/kedacore/governance/blob/main/SCALERS.md)
- [x] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [x] Tests have been added
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [x] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

<!--
  Make sure to link the related issue for this change
  If it requires multiple PRs and/or a PR on another repo as well, please use "Relates to" instead.
-->
Fixes #5067 

<!--
  Make sure to link the related PRs for changes such as documentation & Helm charts
-->
Relates to https://github.com/kedacore/charts/pull/547 and https://github.com/kedacore/keda-docs/pull/1242

zroubalik reviewed (commented):
Looking good! 

@loispostula do you think you can add unit test for this? `hashicorpvault_handler.go` is missing test coverage, so we should work on improving this 🙏

loispostula commented:
@zroubalik I'm working on adding a e2e test as well, and while adding it I discover an issue, so I'll update my commit with all that

loispostula commented:
@zroubalik I've changed the approach for the PR.

I've added support for different vault secret backend. The default behaviour remains the same, if no type is probided, the secret is handled as either a v1 or v2 secret (to keep behaviour of #4152). This allows me to conduct specific `pki` related unmarshalling as not all keys are string for that one. This would allow to specified additional type in the future if needs be.

Regarding testing, I've move the secret fetching inside of the `HashicorpVaultHandler` struct, this allows me to make it easier to test the functionalities of it.

I've also made another notable change which is to bundle the call to the vault api by path. This is the default behaviour of `consul-template` and is especially useful in dynamic secret, as you only want to fetch the secret once and access the different keys

zroubalik commented on apis/keda/v1alpha1/triggerauthentication_types.go:
```diff
@@ -197,11 +197,34 @@ const (
+	VaultSecretTypeSecretV2 VaultSecretType = "secret_v2"
+	VaultSecretTypeSecret   VaultSecretType = "secret"
+	VaultSecretTypePki      VaultSecretType = "pki"
+)
+
+type VaultPkiData struct {
+	CommonName string `json:"common_name,omitempty"`
+	AltNames   string `json:"alt_names,omitempty"`
+	IPSans     string `json:"ip_sans,omitempty"`
+	URISans    string `json:"uri_sans,omitempty"`
+	OtherSans  string `json:"other_sans,omitempty"`
+	TTL        string `json:"ttl,omitempty"`
+	Format     string `json:"format,omitempty"`
+}
+
 // VaultSecret defines the mapping between the path of the secret in Vault to the parameter
 type VaultSecret struct {
-	Parameter string `json:"parameter"`
-	Path      string `json:"path"`
-	Key       string `json:"key"`
+	Parameter string          `json:"parameter"`
+	Path      string          `json:"path"`
+	Key       string          `json:"key"`
+	Type      VaultSecretType `json:"type,omitempty"`
+	PkiData   VaultPkiData    `json:"pki_data,omitempty"`
+	Value     string          `json:"-"`
```
My only concern here is that we generally use Camel Case notation in our resources, not Snake Case. I would prefer if we keep it consistent.

zroubalik commented on pkg/scaling/resolver/hashicorpvault_handler_test.go:
```diff
@@ -0,0 +1,379 @@
+/*
+Copyright 2022 The KEDA Authors
```
```suggestion
Copyright 2023 The KEDA Authors
```

nit

zroubalik commented on pkg/scaling/resolver/hashicorpvault_handler.go:
```diff
@@ -170,9 +175,163 @@ func (vh *HashicorpVaultHandler) Read(path string) (*vaultapi.Secret, error) {
 	return vh.client.Logical().Read(path)
 }
 
+func (vh *HashicorpVaultHandler) Write(path string, data map[string]interface{}) (*vaultapi.Secret, error) {
```
Would be nice, if we can add documentation for funcitons defined in this file. With a short description on what it is doing. Thanks!

zroubalik reviewed (commented):
Thanks a lot for the great contribution! I like the general direction, I love the tests! I have just a few nits.

zroubalik commented:
/run-e2e hashicorp

zroubalik commented:
> /run-e2e hashicorp **Update:** You can check the progress [here](https://github.com/kedacore/keda/actions/runs/6496124047)

@loispostula  what I can see in the test logs logs:

```
Execution of tests/secret-providers/hashicorp_vault/hashicorp_vault_test.go, has failed after "three" attempts
```

```
***0***3-***0-***T***3:36:05Z	ERROR	failed to ensure HPA is correctly created for ScaledObject	***"controller": "scaledobject", "controllerGroup": "keda.sh", "controllerKind": "ScaledObject", "ScaledObject": ***"name":"hashicorp-vault-test-so","namespace":"hashicorp-vault-test-ns"***, "namespace": "hashicorp-vault-test-ns", "name": "hashicorp-vault-test-so", "reconcileID": "30b589***a-88d0-4dfa-b***5d-df9adfd0dd86", "error": "error par sing prometheus metadata: no cert given"***
```

loispostula commented:
> > /run-e2e hashicorp **Update:** You can check the progress [here](https://github.com/kedacore/keda/actions/runs/6496124047)
> 
> @loispostula what I can see in the test logs logs:
> 
> ```
> Execution of tests/secret-providers/hashicorp_vault/hashicorp_vault_test.go, has failed after "three" attempts
> ```
> 
> ```
> ***0***3-***0-***T***3:36:05Z	ERROR	failed to ensure HPA is correctly created for ScaledObject	***"controller": "scaledobject", "controllerGroup": "keda.sh", "controllerKind": "ScaledObject", "ScaledObject": ***"name":"hashicorp-vault-test-so","namespace":"hashicorp-vault-test-ns"***, "namespace": "hashicorp-vault-test-ns", "name": "hashicorp-vault-test-so", "reconcileID": "30b589***a-88d0-4dfa-b***5d-df9adfd0dd86", "error": "error par sing prometheus metadata: no cert given"***
> ```

Thanks a lot, I've replicated locally, and I had not update the e2e test to use the `camelCase`, retriggering a local test then I'll push

zroubalik commented:
/run-e2e hashicorp

zroubalik reviewed (approved):
LGTM, great job! Thanks a lot for this great contribution. 🚀 

Please just resolve the last few things on the docs PR and we can go ahead and merge this!

🙏

loispostula commented:
@zroubalik Thanks, doc pr updated :D Thanks for all the feedback
